#### Goals

 - load a dataset
 - look for distributions
 - explore the data alone different dimensions

## Load the Wine Dataset - (no labels)



In [1]:
var fs = require("fs");
var parse = require('csv-parse/lib/sync');
var table = require('text-table');
var Plot = require('plotly-notebook-js');
var groupBy = require('lodash/groupBy');
var keys = require('lodash/keys');

'use strict'

In [2]:
var raw_csv_string = fs.readFileSync("datasets/wine.data.unlabelled.csv").toString();
var dataset = parse(raw_csv_string);
console.log("our dataset has", dataset.length, "rows and ", dataset[0].length, "columns")

our dataset has 178 rows and  13 columns


In other words we have a dataset cointaining 178 feature vectors, each made up of 13 features. Each feature vector corresponds to a particular wine.

So we have a 13-dimensional data space that we could potentially do our analysis in. That's $R^{13}$!

In [3]:
var features = [
    'Alcohol',
    'Malic Acid',
    'Ash',
    'Alcalinity of ash',
    'Magnesium',
    'Total phenols',
    'Flavanoids',
    'Nonflavanoid phenols',
    'Proanthocyanins',
    'Color intensity',
    'Hue',
    'OD280/OD315 of diluted wines',
    'Proline'];

var format = { align: [ 'c', 'c' ], hsep: ' | ' };

var T = table([features], format);
console.log(T)
console.log("---------------------------")
var D = table(dataset.slice(0,5), format);
console.log(D)

Alcohol | Malic Acid | Ash | Alcalinity of ash | Magnesium | Total phenols | Flavanoids | Nonflavanoid phenols | Proanthocyanins | Color intensity | Hue | OD280/OD315 of diluted wines | Proline
---------------------------
14.23 | 1.71 | 2.43 | 15.6 | 127 | 2.8  | 3.06 | 0.28 | 2.29 | 5.64 | 1.04 | 3.92 | 1065
 13.2 | 1.78 | 2.14 | 11.2 | 100 | 2.65 | 2.76 | 0.26 | 1.28 | 4.38 | 1.05 | 3.4  | 1050
13.16 | 2.36 | 2.67 | 18.6 | 101 | 2.8  | 3.24 | 0.3  | 2.81 | 5.68 | 1.03 | 3.17 | 1185
14.37 | 1.95 | 2.5  | 16.8 | 113 | 3.85 | 3.49 | 0.24 | 2.18 | 7.8  | 0.86 | 3.45 | 1480
13.24 | 2.59 | 2.87 | 21   | 118 | 2.8  | 2.69 | 0.39 | 1.82 | 4.32 | 1.04 | 2.93 | 735


Each of those feature is some measurement made on a wine. We have wine data from a number of different geographical regions.

The point of this dataset is can we work out which wines are from teh same region? Can we even work out how many different regiosn we have wines from?

Algorithms aside, how would you do that?

## Lets plot some data

### 1D

We could plot a simple histogram of any dimension we choose. Can we determine the number of regions or assign any wine to a region from a single dimension?

In [ ]:
var decimals = 1;
var factor = Math.pow(10,decimals);

// select a feature
var DIM = 0;
// -------
var groups = groupBy(dataset.map((row) => row[DIM]), v => Math.floor(factor*v)/factor);
var values = keys(groups);
var histogram = values.map(value => groups[value].length);

var trace = { x: values.map(s => parseFloat(s)), y: histogram, type: 'bar', name: 'cosine'}
var layout = { title: `Feature: ${features[DIM]}`, xaxis: { title: features[DIM] }, yaxis: { title: "Value" }};

var myPlot = Plot.createPlot([trace], layout);

$$html$$ = myPlot.render();

### Move to 2D


In [ ]:
var X = 0;
var Y = 11;

var x = dataset.map(d => d[X]).map(f => parseFloat(f));
var y = dataset.map(d => d[Y]).map(f => parseFloat(f));

var singleTrace = { x, y, mode: 'markers', marker: { size: 5, }, type: 'scatter' };

var layout = { xaxis: { title: features[X] }, yaxis: { title: features[Y] }, width: 500, height: 500 };

$$html$$ = Plot.createPlot([singleTrace], layout).render()

#### What about in 3D or 4D?

In [ ]:
var X = 0, Y = 11, COLOR = 6, SIZE = 1;

var x = dataset.map(d => d[X]).map(f => parseFloat(f));
var y = dataset.map(d => d[Y]).map(f => parseFloat(f));
var color = dataset.map(d => d[COLOR]).map(f => parseFloat(f));
var size = 10; //dataset.map(d => d[SIZE]).map(f => parseFloat(f));
var singleTrace = { x, y, mode: 'markers', marker: { color, size, sizeref: 0.2 }, type: 'scatter' };
var layout = { title: `Size ${features[SIZE]} | Color: ${features[COLOR]}`, 
  xaxis: { title: features[X] }, yaxis: { title: features[Y] }, width: 700, height: 700 };

$$html$$ = Plot.createPlot([singleTrace], layout).render()